ls1intum/Artemis
# `Programming exercise`: Use the git-diff to show line stats

### Checklist
#### General


#### Client

### Motivation and Context


Since we now generate the git-diff report for every exercise it makes sense to use this information in other ways.
One way to utilize this is to show the amount of work that the students have to do for an exercise.

### Description

The amount of work can be roughly described by the number of lines the students have to add/change. The simplest way to show this is a line stat view that shows the added and removed lines from the git-diff. This line stat view was inspired by the one GitHub uses.
The line-stat view can be seen in 3 places:
1. At the bottom of the detail view (if the git-diff report exists)
3. At the top right of the git-diff report modal
4. At the top right of each git-diff entry in the modal
    - This one is a bit temporary, as it will be reworked when we improve the diff view

One thing to note here is that 2 and 3 represent filtered line stats meaning that they filter out empty lines, as these don't really represent any amount of work. 1 on the other hand represents an unfiltered line stat, as we do not have access to the full git-diff report (with the actual code) here.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise

If you encounter an internal server error when creating new exercises see #4893 

1. Log in to Artemis
5. Go to the exercise detail page
7. Open the git diff modal
8. Check that the filtered line counts are correct
9. Close the modal
6. Look at the bottom if the unfiltered line count is correct

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| New client code | 100% | 100% |

### Screenshots


The git-diff modal: 
<img width="1133" alt="grafik" src="https://user-images.githubusercontent.com/18218285/162064355-368a5f65-a185-456d-8ca3-89cb2723cc01.png">
The detail page:
<img width="404" alt="grafik" src="https://user-images.githubusercontent.com/18218285/162064390-753864fb-2353-4eee-ba4e-81f5a04ecd0e.png">
